Dataloader changes for RLVR

gupta-abhay opened the pull request:
- [x] Add script changes to write data
- [x] Add dataloader changes for verified answer
- [x] Add verifiable rewards (both defn and to registry)
- [x] Handle `str` vars for dataloader and callbacks, enabling path to more verifiers
- [x] Move verifier code to a single `rlvr_utils` file

vchiley reviewed (changes requested):
Going to channel my inner Daniel: "add tests."

SeanKski commented on scripts/data/unified_tokenize_dataset.py:
```diff
@@ -128,6 +144,49 @@ def _process_classifier_sample(self, sample: Any):
+
+        This function is currently hard-coded for the GSM8K dataset.
+        """
+        prompt_fn = prepare_gsm8k_prompt
```
maybe throw a log.warning here saying this this is hard-coded for GSM8K  (just to avoid a gotchya)

SeanKski reviewed (commented):
Overall, I think this looks good. 

One quality-of-life thing that I think we should implement (but is not necessary now, so they shouldn't block this merge) is: 

  Rather than having an additional `<EvalName>FormatVerificationReward`, `<EvalName>AnswerVerificationReward` for every eval, we should build out some abstractions/standardizations, such as, a base `FormatVerificationReward` and `AnswerVerificationReward` which we then subclass/configure for the different evals. 
This way we won't overload the main `reward_learning` registry and cut down on repeated code 
e.g., otherwise, each outcome reward will have its own version of: 
```python 
       for i in range(batch_size):
            _answer = self._extract_solution(all_generated_texts[i])
            _reward = self._score_generations(_answer, verified_answers[i])
            rewards[i, generated_lens[i] - 1] += _reward
        return rewards
```

gupta-abhay commented:
@SeanKski - addressed all your comments
@vchiley - added tests for rewards and for the dataloader workflow
@dakinggg - adding more dependencies - ptal if they look good - specifically the omegaconf installs which is required for antlr workflows.

dakinggg commented on pyproject.toml:
```diff
@@ -15,6 +15,10 @@ dependencies = [
     'kubernetes==29.0.0',
     'vllm>=0.8.2,<0.9.0',
+    # for MATH dataset
+    'omegaconf==2.4.0.dev3',
```
can you explain why you need exact pins for all of these libraries?

dakinggg replied:
I skimmed the omegaconf commits between 2.3 and 2.4.0.dev3 and don't have a concern about that one. 

can you do a >=1.13.1,<2 for sympy?

where does the antlr dep come from? i.e. who uses it and do they have the dep pinned?

gupta-abhay replied:
- The `antlr-runtime` and `sympy` library are sort of intertwined - they need to be fixed to versions, else it leads to weird parsing errors. 
- For `omegaconf`, if I do not specify the library, it defaults to the `2.3.0` - which does not suport `antlr-4.11` -- which is inturn needed for the parsing.

dakinggg replied:
ok i see, there is a dynamic version check in sympy when you call the parse latex stuff. Can you leave a comment behind explaining these things?

dakinggg replied:
inline here in pyproject.toml

vchiley commented on compose_rl/utils/utils.py:
```diff
@@ -1084,3 +1085,12 @@ def make_action_mask(
     return mask
+
+
+def flatten(coll: Union[Iterable[Any], str]) -> Generator[Any, None, None]:
```
doc str
(cuz i'm pedantic)
use this one or do you

```suggestion
def flatten(coll: Union[Iterable[Any], str]) -> Generator[Any, None, None]:
    """Recursively flattens an arbitrarily nested iterable (excluding strings).

    Note: strings are treated as atomic elements and are not flattened into
    characters.

    Args:
        coll (Union[Iterable[Any], str]): The nested iterable to flatten.

    Yields:
        Any: The individual, non-iterable elements from the flattened structure.
    """
```

abaheti95 reviewed (approved):
Looks good. Thanks!
Probably need to remove one function that is unused.

gupta-abhay commented on compose_rl/data/rlvr_utils.py:
```diff
@@ -0,0 +1,266 @@
+        return False
+
+
+def get_unnormalized_answer(text: str) -> str:
```
ahh good catch, will remove

dakinggg commented on compose_rl/data/prompt_data.py:
```diff
@@ -90,5 +98,19 @@ def __getitem__(self, idx: int) -> dict[str, Any]:
+                _answer = verified_answer
+            else:
+                try:
+                    _answer = verified_answer.decode('utf-8', errors='strict')
```
😱 you have verified answers with non utf-8 decodable

gupta-abhay replied:
extra checks if any user makes weird datasets ❌

dakinggg commented on compose_rl/data/prompt_data.py:
```diff
@@ -90,5 +98,19 @@ def __getitem__(self, idx: int) -> dict[str, Any]:
+            else:
+                try:
+                    _answer = verified_answer.decode('utf-8', errors='strict')
+                except UnicodeDecodeError:
+                    print(f'Failed to decode verifed_answer')
```
nit: log instead of print

vchiley replied:
```suggestion
                except UnicodeDecodeError as e:
                    log.error(f'Failed to decode verifed_answer with error: {e}')
```

dakinggg commented on compose_rl/data/rlvr_utils.py:
```diff
@@ -0,0 +1,252 @@
+    return s.strip('{}')
+
+
+class timeout:  # noqa: N801
```
lol i asked claude whether this approach or the one in foundry (https://github.com/mosaicml/llm-foundry/blob/60e24b066d0037678de54c5ffaaf09b91ea9d8b7/llmfoundry/callbacks/run_timeout_callback.py#L4) works better, and the answer was that the one in foundry works on windows 😂

dakinggg replied:
I suppose the one in foundry (threading.Timer) might be slightly better cause the signal.alarm calls here are process global right?

gupta-abhay replied:
yeah, you might be right 🤔 . For now, I will use this, since the minieval class uses the same functions - helps keep the evals and training closer to each other. We can mark this for future improvements as needed.

vchiley commented on compose_rl/data/prompt_data.py:
```diff
@@ -90,5 +98,19 @@ def __getitem__(self, idx: int) -> dict[str, Any]:
+        item_dict = {'prompt': prompt, 'prompt_len': prompt_len}
+
+        verified_answer = sample.get('verified_answer', None)
+        if verified_answer:
```
```suggestion
        if verified_answer is not None:
```
in case verified answer is an empty string (`''`) for some reason
or should this case be caught as a separate case and dealt with separately?

gupta-abhay replied:
this is handled in the initial data processing script here: https://github.com/databricks/compose-rl/blob/abhay/data_loader_rlvr/scripts/data/unified_tokenize_dataset.py#L208

gupta-abhay replied:
so we will not have empty strings here

vchiley commented on compose_rl/reward_learning/functional.py:
```diff
@@ -242,3 +250,190 @@ def __call__(
+            f'Using reward value of {self.reward} for {self.__class__.__name__} verifier',
+        )
+
+    def validate_config(self):
+        # Base validation, can be overridden
+        return
```
@bcui-db should validate_config be a member method of `Reward` cls since a bunch of these fn's define it?


Also
`pass` == `return` == `return None`, but `pass` is slightly more explicit for indicating a deliberate no-operation placeholder.
```suggestion
    def validate_config(self):
        # Base validation, can be overridden
        pass
```
maybe I'm just pedantic 🤷‍♂️

gupta-abhay replied:
@vchiley - noted. Will raise a separate PR for this. This Pr is already pretty large, and this will affect several files.

vchiley commented on compose_rl/reward_learning/functional.py:
```diff
@@ -242,3 +250,190 @@ def __call__(
+        Note: The label parameter is not used in this implementation but is required
+        by the interface.
+        """
+        solution = re.search(r'####.*?([\d,]+(?:\.\d+)?)', answer)
```
does this line belong in the `extract_solution` fn?
single line fn

gupta-abhay replied:
can be done that way - but since we just need to check if the 4 hashes are followed, this simplifies stuff by putting it in a single function. else, it will complicate the interface with the `extract_solution` needing more fine grained handling of a None input. I dont think we currently have that diversified a need, and can let this be here.

vchiley commented on scripts/data/unified_tokenize_dataset.py:
```diff
@@ -31,19 +39,25 @@ def __init__(
 
         print(f'Dataset name: {dataset_name}')
+        if subset:
+            print(f'Processing subset: {subset}')
```
this file has a bunch of prints
can you update these to `log` instead
🙏

vchiley commented on tests/common/datasets.py:
```diff
@@ -72,3 +72,20 @@ def __getitem__(self, index: int):
+    def __getitem__(self, index: int):
+        return {
+            'prompt': torch.ones((self.prompt_len,)).int(),
+            'prompt_len': torch.Tensor([self.prompt_len]).to(torch.int64),
```
torch.int64 needed?
why not `uint32`
goes from 0 to > 4B?

gupta-abhay replied:
Keeps this the same as the actual code definition - which does int64 -- we can mark for future refinements as needed.

vchiley reviewed (approved):
left a few pedantic comments
but generally, lgtm

gupta-abhay commented on compose_rl/reward_learning/functional.py:
```diff
@@ -242,3 +250,190 @@ def __call__(
+        Returns:
+            float: The reward score.
+        """
+        pass
```
since this is an abstractmethod - pass should be fine. Had it not been an abstract method - would necessitate the `raise` method.
